### DISCUSSION TITLE
io.smallrye.graphql.client.InvalidResponseException

### DISCUSSION THREAD

**Discussion by maag03** [quarkusio/quarkus#29181] [Q&A]
Using the smallrye graphql dynamic client and moving from Quarkus v2.9.2.Final to v2.10.3.Final (and later) we get this exception upon receiving the response from remote:
io.smallrye.graphql.client.InvalidResponseException: Unexpected response. Code=200, message="", body="{"data":{"CustomerMandates":{"select":[],"total":0}},"errors":[],"dataPresent":true,"extensions":null}"

Any idea what issue could be ?

Regards
Magnus

[a comment by an automated account is not shown]

**Comment by jmartisk** [comment]
To me this looks like the response body is not a well-formed JSON document. One closing curly brace is missing. Is the service actually responding with that, or perhaps we are losing the last character from the response for some reason?

**Comment by maag03** [comment]
This is how the actual response looks like:
{
    "data": {
        "CustomerMandates": {
            "select": [],
            "total": 0
        }
    },
    "errors": [],
    "dataPresent": true,
    "extensions": null
}

**Reply by jmartisk** [reply]
Ah right, sorry, I was looking at it wrong.
The problem seems to be that the body contains a field named `dataPresent`, which is forbidden per the GraphQL specification. The only allowed fields are `data`, `errors` and `extensions`.

See http://spec.graphql.org/draft/#sec-Response-Format: 
> To ensure future changes to the protocol do not break existing services and clients, the top level response map must not contain any entries other than the three described above.

The GraphQL client chooses to throw an error in such case, because the response contains items that we don't know how to handle. We can't pass it to the application code, because the `Response` class doesn't have any appropriate fields for such unexpected items.

Do you have control over the server that returns this? If yes, I would suggest fixing that on the server side.
If not, then... hmm, we might be able to add a configuration flag that allows to ignore such fields (log a warning) instead of failing. But I think we should keep failing as the default behavior.

**Comment by maag03** [comment]
Thx for reply Jan! We'll speak to server side owners and see if they can correct this.
_
Magnus

**Reply by jmartisk** [reply]
Did you find out anything? :)

**Reply by jmartisk** [reply]
In any case I've reported https://github.com/smallrye/smallrye-graphql/issues/1650 because I think it could be useful for other people too, so in a future version we might implement it

**Reply by maag03** [reply]
They will probably remove the invalid field from response.
Yes, I think that could be useful.
